### PR TITLE
Mplayer improvements

### LINKED ISSUE #130: Display the score for all Landers in that session (co-op)

### LINKED ISSUE #191: Pressing R will delete all clients when it shouldn't

### LINKED ISSUE #192: Client connecting to host has delay. See comments.
Client takes a long time to connect. Need a status message or something on client so they have feedback.

### LINKED ISSUE #193: Need to present host with message when client joins

### LINKED ISSUE #227: Don't send MASS data over enet. See comments.
Try sending a skinny version of the lander table across enet and let the client determine draw as much as possible on the client side.

### LINKED ISSUE #242: Save/load game should probably be disabled in mplayer mode

### PR DESCRIPTION
**Pull request by togfoxy**
Lots of little mplayer nice things and one or two bug fixes.

Fixes #191 
Fixes #192 
Fixes #193 
Fixes #242 
Fixes #130 
Fixes #227

### CODE REVIEW

**Comment by togfoxy** [comment]
Okay. I tested this one. I think it's good. Please review and merge when ready. Thanks.

File: source/functions.lua [hunk 1]
```diff
@@ -202,73 +203,24 @@ function functions.GetDistanceToClosestBase(xvalue, intBaseType)
...
+	-- Smoke.destroy()
 
 	-- ensure Terrain.init appears before Lander.create
-	Terrain.init()
-
-	LANDERS = {}
-	table.insert(LANDERS, Lander.create())
-
+	-- Terrain.init()
+
```
Comment (MadByteDE) [inline source/functions.lua:216]: Is this meant to be commented out?

File: source/main.lua [hunk 2]
```diff
@@ -180,6 +180,10 @@ end
 
 function love.load()
     if love.filesystem.isFused() then
+	
+		-- nullify the assert function for performance reasons
+		function assert() end
+
```
Comment (MadByteDE) [inline source/main.lua:186]: Can you explain what it does regarding performance? I never experienced a big impact in performance using assert to check statements and variables.

File: CHANGELOG.md [hunk 3]
```diff
@@ -34,7 +36,8 @@ The file format is based on [Keep a Change Log](https://keepachangelog.com/en/1.
 - loading a game that doesn't exist no longer crashes the game
 - rangefinder fixed
 - fixed the '$' sign in the shop
-- Fix smoke particles not being removed after starting a new game
+- fix smoke particles not being removed after starting a new game
```
Comment (MadByteDE) [inline CHANGELOG.md:39]: Since Smoke.destroy is commented out the particles won't get removed when starting a new game.

**Review by MadByteDE — CHANGES_REQUESTED** [review]
Nice, disabling saving/loading is a valid workaround. 
What I miss for the MP is the ability to disconnect from the server and go back to SP without restarting the whole game.

**Some more issues:**
- Smoke still will be present upon game reset
- Clients won't get removed from the MP session when leaving / disconnecting. This applies to the server side as well as the client side.

**Video:**

https://user-images.githubusercontent.com/4752038/141611828-63775309-5dfe-4988-baaf-f92661c5ab12.mp4

**Comment by togfoxy** [comment]
I restored those lines and will work on them in a later PR.
- there is no need to regenerate the ground table
- there is no need to regenerate the objects table but they do need to be reset with fuel and money
- the smoke particles can play out to the end of the animation then disappear. It looks pretty cool.

That's not the focus of this PR so I removed them.

Nullifying the assert is just good coding practice and frankly, don't want them to trigger a runtime error in a released version.

**Comment by togfoxy** [comment]
I'll disable the multiple joining thing on the next iteration. Never thought anyone would try that.  :)

**Comment by MadByteDE** [comment]
It also seems like the buildings are not generated deterministic for server and clients as seen here:

[](https://i.imgur.com/8q7vEJf.png)

**Comment by togfoxy** [comment]
> It also seems like the buildings are not generated deterministic for server and clients as seen here: 8q7vEJf

Correct. They are random for now. They will need to be generated in a formula at some point so they won't need to be exchanged across enet in multiplayer mode.

**Pull request merged**